src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http2Connection.cs

```diff
@@ -1285,6 +1335,15 @@ public sealed override async Task<HttpResponseMessage> SendAsync(HttpRequestMess
                     // ISSUE 31315: Determine if/how to expose HTTP2 error codes
                     throw new HttpRequestException(SR.net_http_client_execution_error, e);
                 }
+                else if (e is OperationCanceledException)
+                {
+                    // If the operation has been canceled after the stream was allocated an ID, send a RST_STREAM.
+                    if (http2Stream != null && http2Stream.StreamId != 0)
+                    {
+                        http2Stream.Cancel();
+                    }
+                    throw new OperationCanceledException(cancellationToken);
```

Comment:
Should we instead just propagate the original exception?
```C#
throw;
```
Or is it possible the original exception might not be associated with this cancellation token?  In which case:
```C#
else if (e is OperationCanceledException oce)
{
    ...

    if (oce.CancellationToken == cancellationToken)
    {
        throw;
    }

    throw new OperationCanceledException(cancellationToken);
}
```
?

Reply:
I've tried to be consistent about associated the CancellationToken, but there are some places where I'm depending on other components throwing. I'll opt for the second approach, just to be safe.

Reply:
This actually surfaced a different issue: TrySetCancelled throws TaskCanceledException, but I've setting things up to throw and expect OperationCanceledException. It isn't totally clear to me which we should be throwing, as our behavior throughout SocketsHttpHandler doesn't seem to be consistent. Any thoughts on this?

Reply:
TaskCanceledException derives from OperationCanceledException... it's just an OperationCanceledException that also carries with it the Task that was canceled.  In general developers should be catching the base OCE, not the derived TCE.  However, we've tried to always throw TCE whenever it's appropriate: https://github.com/dotnet/corefx/pull/28075

Reply:
Okay. I'll update this to follow the same pattern.